core/src/test/java/kafka/test/api/ShareConsumerTest.java

```diff
@@ -455,6 +458,49 @@ public void testExplicitAcknowledgeCommitSuccess() {
         shareConsumer.close();
     }
 
+    @Test
+    public void testExplicitAcknowledgementCommitAsync() throws InterruptedException {
+        ProducerRecord<byte[], byte[]> record1 = new ProducerRecord<>(tp.topic(), tp.partition(), null, "key".getBytes(), "value".getBytes());
+        ProducerRecord<byte[], byte[]> record2 = new ProducerRecord<>(tp.topic(), tp.partition(), null, "key".getBytes(), "value".getBytes());
+        ProducerRecord<byte[], byte[]> record3 = new ProducerRecord<>(tp.topic(), tp.partition(), null, "key".getBytes(), "value".getBytes());
+        KafkaProducer<byte[], byte[]> producer = createProducer(new ByteArraySerializer(), new ByteArraySerializer());
+        producer.send(record1);
+        producer.send(record2);
+        producer.send(record3);
+
+        KafkaShareConsumer<byte[], byte[]> shareConsumer1 = createShareConsumer(new ByteArrayDeserializer(), new ByteArrayDeserializer(), "group1");
+        KafkaShareConsumer<byte[], byte[]> shareConsumer2 = createShareConsumer(new ByteArrayDeserializer(), new ByteArrayDeserializer(), "group1");
+        shareConsumer1.subscribe(Collections.singleton(tp.topic()));
+        shareConsumer2.subscribe(Collections.singleton(tp.topic()));
+
+        Map<TopicPartition, Exception> partitionExceptionMap1 = new HashMap<>();
+
+        shareConsumer1.setAcknowledgementCommitCallback(new TestableAcknowledgeCommitCallback(partitionExceptionMap1));
+
+        ConsumerRecords<byte[], byte[]> records = shareConsumer1.poll(Duration.ofMillis(5000));
+        assertEquals(3, records.count());
+        Iterator<ConsumerRecord<byte[], byte[]>> iterator = records.iterator();
+
+        // Acknowledging 2 out of the 3 records received via commitAsync.
+        shareConsumer1.acknowledge(iterator.next());
+        shareConsumer1.acknowledge(iterator.next());
+        shareConsumer1.commitAsync();
+
+        // Allowing acquisition lock timeout to expire.
+        Thread.sleep(10000);
+
+        // The 3rd record should be reassigned to 2nd consumer when it polls.
+        ConsumerRecords<byte[], byte[]> records2 = shareConsumer2.poll(Duration.ofMillis(5000));
+        assertEquals(1, records2.count());
+
+        assertFalse(partitionExceptionMap1.containsKey(tp));
+        // The callback will receive the acknowledgement responses after the next poll.
+        shareConsumer1.poll(Duration.ofMillis(5000));
+
+        assertTrue(partitionExceptionMap1.containsKey(tp));
+        assertNull(partitionExceptionMap1.get(tp));
```

Comment:
Let's close the producer and consumers at the end of the test too. I know that's not done everywhere, but it should be really.